React/calendar react

crossan007 opened the pull request:
#### What's this PR do?
*  Converts the modal dialogue in the Calendar page to use a ReactJS / TypeScript / JSX application instead of the old jQuery app.
*  Adds support for editing existing events directly from the calendar page.
*  Ensures all event edits are made using API calls, and not legacy page post-backs.
*  Adds a few dev dependencies, and a new step to the build process `npm run build-react`
*  Adds support for creating UI forms with JSX, and adds support for TypeScript development
*  Adds support for webpack and babel to generate minified JS / CSS (currently only for React apps)


#### Screenshots (if appropriate)

New View:
![image](https://user-images.githubusercontent.com/11679900/52313720-9214d380-297d-11e9-8f0c-d4724352f207.png)

Old View:
![image](https://user-images.githubusercontent.com/11679900/52313726-980ab480-297d-11e9-8220-39c87f4f79e1.png)


New Edit (there was no old edit view here):
![image](https://user-images.githubusercontent.com/11679900/52313732-a3f67680-297d-11e9-8046-30faa29bdd0d.png)


#### What Issues does it Close?

Closes #4129 
Closes #3725 
Closes #2627


#### What are the relevant tickets?
Unblocks #3723 
Unblocks #2558
Helps with #3036


#### Any background context you want to provide?

#### Where should the reviewer start?

#### How should this be manually tested?

#### How should the automated tests treat this?

#### Questions:
- Is there a related website / article to substantiate / explain this change?
  -  [ ] Yes - Link:
  -  [ ] No

- Does the development wiki need an update?
  -  [ ] Yes - Link:
  -  [ ] No

- Does the user documentation wiki need an update?
  -  [ ] Yes - Link:
  -  [ ] No

- Does this add new dependencies?
  -  [ ] Yes
  -  [ ] No

- Does this need to add new data to the demo database
  -  [ ] Yes
  -  [ ] No

crossan007 commented:
I still need to do a pass through here to make sure I clean up all the unused code

crossan007 commented:
That was a quick review hahaha.  

This is the first time I've added react into an existing project, so maybe a bit more scrutiny would be good ;-)

crossan007 commented:
Still need to figure out how to test the `react-select` components with Selenium: https://github.com/JedWatson/react-select/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+selenium

DawoudIO commented:
lots of changes, can you break out the PHP code into a new PR as it is easier to review ... also maybe break out the code that is not for the calendar into a new PR ...

crossan007 commented:
@DawoudIO I removed the changes to the Vagrantfile and boot strap process.  I also removed dead code pertaining to the calendar and some debugging code that was still in the PR.

At this point, all code in this PR is required for functionality of the new React calendar.

I made a few changes to the names of the properties used in the JSON object representing an event, so I can't really separate out the PHP changes from the JavaScript changes without breaking something.

Also, there are updates included here for the test suites so that the react components can be tested in CI.

crossan007 commented:
Live stream of working on this PR: https://www.youtube.com/watch?v=js4E7IAEbvY&feature=youtu.be
